src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs

```diff
@@ -74,20 +74,15 @@ public DirectoryInfo CreateSubdirectory(string path)
 
             string newPath = Path.GetFullPath(Path.Combine(FullPath, path));
 
-            ReadOnlySpan<char> trimmedNewPath = Path.TrimEndingDirectorySeparator(newPath.AsSpan());
-            ReadOnlySpan<char> trimmedCurrentPath = Path.TrimEndingDirectorySeparator(FullPath.AsSpan());
-
             // We want to make sure the requested directory is actually under the subdirectory.
-            if (trimmedNewPath.StartsWith(trimmedCurrentPath, PathInternal.StringComparison)
-                // Allow the exact same path, but prevent allowing "..\FooBar" through when the directory is "Foo"
-                && ((trimmedNewPath.Length == trimmedCurrentPath.Length) || PathInternal.IsDirectorySeparator(newPath[trimmedCurrentPath.Length])))
-            {
-                FileSystem.CreateDirectory(newPath);
-                return new DirectoryInfo(newPath);
-            }
-
-            // We weren't nested
-            throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, path, FullPath), nameof(path));
+            // Allow the exact same path, but prevent allowing "..\FooBar" through when the directory is "Foo"
+            string newPathWithTrailingSeparator = PathInternal.EnsureTrailingSeparator(newPath);
+            string currentPathWithTrailingSeparator = PathInternal.EnsureTrailingSeparator(FullPath);
+            if (!newPathWithTrailingSeparator.StartsWith(currentPathWithTrailingSeparator, PathInternal.StringComparison))
+                throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, newPath, FullPath), nameof(path));
```

Comment:
```suggestion
            if (!newPathWithTrailingSeparator.StartsWith(currentPathWithTrailingSeparator, PathInternal.StringComparison))
            {
                throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, newPath, FullPath), nameof(path));
            }
```

Reply:
This is perfectly fine with me, but out of curiosity, what's the reason we'd add braces around this one when all the other `if(X) throw Y;` lines just above this just use indentation for the single-statement block? I did read the code style guidelines for the project, but I guess I must have missed something there.

Reply:
We could use https://github.com/dotnet/runtime/blob/72e41f9c89a0874352ad013dd0534e986a8819dc/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/InMemoryDirectoryInfo.cs#L120-L127

Reply:
@iSazonov - Thanks for the suggestion. How are you proposing we put that function to use? The solution I've proposed up to this point doesn't actually do a special-case check for being a root directory at all; instead, it adjusts the logic for one of the forbidden cases so that it doesn't accidentally also disallow creating subdirectories of the root directory.

Reply:
> doesn't actually do a special-case check for being a root directory at all

The comment in line 77 suggests otherwise. Actually FullPath must be a root path for newPath.

Reply:
Oh, when you say "root path" do you mean "parent path" of another path? Because "IsRootDirectory" to me is a method name that suggests it checks to see if a path is an actual root directory, an absolute root, as in "/" or "C:\" or the like. Especially since this bug we're addressing here is specifically a failure in the case where it is used on a drive root directory, I was thrown off by the ambiguity of the term "root". (In fact, I'd suggest maybe the above method should be called "IsParentDirectory" instead.)

In any case, I'll take a look at that- need to make sure using it doesn't regress what was being fixed by 6c41cc9f889d3faa1ed03a8906632a0a5ad57b23.

Reply:
I agree that the method name confuses.